### PR TITLE
2.0 cleanup and advanced stats

### PR DESCRIPTION
**Pull request by RdN**
Cleaning up 2.0 version:
-cleaning up .gitignore file
-cleaning up vendors - no sense to keep them in repo - we have composer.lock
-forcing to disable debug while running phpunit tests
-#87 adding more location stats (datetime, ip, country, referer)

### CODE REVIEW

**Comment by cydrobolt** [comment]
Thanks for making this massive PR! Will test and get back to you soon :)

**Comment by cydrobolt** [comment]
Regarding vendor files, is there a reason why you've removed them in your PR? We like to keep the vendored files in the repo in case authors decide to delete their projects or if a user would like to use Polr in an environment without `composer` or if the user is not familiar with `composer`.

File: .gitignore [hunk 1]
```diff
@@ -1,8 +1,7 @@
-#/vendor
+vendor/
 bootstrap/cache/
 storage/
-env
+.env.*.php
```
Comment (cydrobolt) [inline .gitignore]: This shouldn't be ignored. 
`.env.blade.php` is the configuration template used by the setup script, which should remain in the Git repo and not ignored.

File: .gitignore [hunk 2]
```diff
@@ -1,8 +1,7 @@
...
+.env.php
```
Comment (cydrobolt) [inline .gitignore]: What's the purpose of this line?

File: app/Http/Controllers/LinkController.php [hunk 3]
```diff
@@ -82,18 +83,19 @@ public function performRedirect(Request $request, $short_url, $secret_key=false)
...
-        $clicks = intval($link->clicks);
+        $referer = $request->server('HTTP_REFERER');
+        $ip = $request->getClientIp();
 
-        if (is_int($clicks)) {
-            $clicks += 1;
-        }
-        $link->clicks = $clicks;
-        $link->save();
+        Click::create([
```
Comment (cydrobolt) [inline app/Http/Controllers/LinkController.php]: More advanced statistics should be an optional feature; we should still continue to increment clicks on the link model whether advanced stats are enabled or not. If advanced stats are disabled, we should avoid inserting rows for each click into the database and hide the advanced features. This will require a  toggle in the setup.

File: resources/views/snippets/link_table.blade.php [hunk 4]
```diff
@@ -26,7 +26,22 @@
...
-        <td>{{$link->clicks}}</td>
+        <td>
+            <div class="btn-group">
+                <button type="button" class="btn btn-sm btn-info"><span class="badge">{{count($link->clicks)}}</span></button>
+                <button type="button" class="btn btn-sm btn-info dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">
+                    <span class="caret"></span>
+                    <span class="sr-only">Toggle Dropdown</span>
+                </button>
+                <ul class="dropdown-menu">
+                    <li><a href="" ng-click="showStats($event, '{{route('stats', ['link_id' => $link->id, 'type' => 'day'])}}')">Clicks by day</a></li>
```
Comment (cydrobolt) [inline resources/views/snippets/link_table.blade.php]: Perhaps it would be more concise to provide a separate page or modal with all the graphs listed together? Separate links are a bit hard to parse.

File: resources/views/snippets/link_table.blade.php [hunk 5]
```diff
@@ -26,7 +26,22 @@
...
+                    <li><a href="" ng-click="showStats($event, '{{route('stats', ['link_id' => $link->id, 'type' => 'country'])}}')">Clicks by country</a></li>
+                    <li><a href="" ng-click="showStats($event, '{{route('stats', ['link_id' => $link->id, 'type' => 'referer'])}}')">Clicks by referer</a></li>
+                    <li role="separator" class="divider"></li>
+                    <li><a href="{{route('clicks', ['link_id' => $link->id])}}">List all</a></li>
```
Comment (cydrobolt) [inline resources/views/snippets/link_table.blade.php]: This could be moved to the header or footer of a modal with all the graphs combined.

File: app/Http/Controllers/StatController.php [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+    /**
+     * @param string $type
+     * @param string $linkId
+     * @return RedirectResponse|View
+     */
+    public function makeReport($type, $linkId)
+    {
+        if (!in_array($type,['day', 'country', 'referer']))
+        {
+            return redirect(route('admin'))->with('error', "Such report does not exist.");
```
Comment (cydrobolt) [inline app/Http/Controllers/StatController.php]: => "such a report does not exist" 
typo?

File: app/Http/Controllers/StatController.php [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+        if ($check = $this->performAuthCheck($link))
+        {
+            return $check;
+        }
+
+        switch ($type) {
+            case 'day':
+                $stats = DB::table('clicks')
+                    ->select(DB::raw("DATE_FORMAT(date, '%Y-%m-%d') as date, count(*) as clicks"))
+                    ->where('link_id', '=', $link->id)
```
Comment (cydrobolt) [inline app/Http/Controllers/StatController.php]: The '=' seems redundant

File: app/Http/Controllers/StatController.php [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+
+        if ($check = $this->performAuthCheck($link))
+        {
+            return $check;
+        }
+
+        switch ($type) {
+            case 'day':
+                $stats = DB::table('clicks')
+                    ->select(DB::raw("DATE_FORMAT(date, '%Y-%m-%d') as date, count(*) as clicks"))
```
Comment (cydrobolt) [inline app/Http/Controllers/StatController.php]: It would be preferable to avoid using raw SQL to perform these queries. Is it possible to use Eloquent's built in functions to retrieve this information rather than performing raw queries? https://laravel.com/docs/5.1/eloquent

File: app/Http/Controllers/StatController.php [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+        return view('stats', ['link' => $link, 'stats' => $stats, 'type' => $type]);
+    }
+
+    /**
+     * @param Link $link
+     * @return RedirectResponse
+     */
+    private function performAuthCheck($link)
+    {
+        if (!$this->isLoggedIn()) {
```
Comment (cydrobolt) [inline app/Http/Controllers/StatController.php]: `self::` should be used here for access to static function `isLoggedIn()`

File: app/Http/Controllers/StatController.php [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+            return redirect(route('login'))->with('error', 'Please login to access your dashboard.');
+        }
+
+        $username = session('username');
+
+        if (!$this->currIsAdmin() && $link->creator != $username)
+        {
+            return redirect(route('admin'))->with('error', "You don't have access to that link stats.");
```
Comment (cydrobolt) [inline app/Http/Controllers/StatController.php]: => "to those link stats."

File: database/migrations/2016_09_24_095751_create_stats_table.php [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+            $table->string('referer_host')->nullable();
+            $table->unsignedInteger('link_id');
+
+            $table->index('ip');
+            $table->index('referer');
+
+            $table->foreign('link_id')->references('id')->on('links')->onDelete('cascade');
+        });
+
+        Schema::table('links', function ($table) {
```
Comment (cydrobolt) [inline database/migrations/2016_09_24_095751_create_stats_table.php]: re: previous comment about optional nature of advanced stats

The "clicks" column should be maintained and advanced stats should be optional. We also want to be as backwards-compatible as possible.

**Comment by RdN** [comment]
@cydrobolt composer.lock is enough - it keeps installed versions. Please check:
https://getcomposer.org/doc/faqs/should-i-commit-the-dependencies-in-my-vendor-directory.md

In case user is not familiar with composer - we should add details, in documentation. Also composer has pretty good documentation.

But still it is up to you.

More important is why this test is failing - I ran it locally (follow travis steps), and it works fine for me. Also if we want to test fetching country code for IP, test environment should have MAxmindDB downloaded (as I described in documentation).

**Comment by cydrobolt** [comment]
@RdN yeah, will definitely be a tough question to answer definitively. I feel as if the user base has changed to more technical users mostly familiar with `composer`, although that was not the case in the past. We can evaluate this in the end :)

I'll likely be revamping the graphs and make them pretty with d3.js and other charting libraries, if needed. Thank you so much for setting up the Clicks system, though! Exactly what I had in mind in terms of the database structure.

File: database/migrations/2016_09_24_095751_create_stats_table.php [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+            $table->timestamp('date');
+            $table->string('ip');
+            $table->string('country')->nullable();
+            $table->string('referer')->nullable();
+            $table->string('referer_host')->nullable();
+            $table->unsignedInteger('link_id');
+
+            $table->index('ip');
+            $table->index('referer');
+
```
Comment (cydrobolt) [inline database/migrations/2016_09_24_095751_create_stats_table.php]: Could we also fetch and store the user agent in these rows?

File: resources/views/layouts/base.blade.php [hunk 13]
```diff
@@ -37,9 +37,14 @@
     <link href="/css/base.css" rel="stylesheet">
     <link href="/css/toastr.min.css" rel="stylesheet">
     <link href="/css/font-awesome.min.css" rel="stylesheet">
+    <link rel="stylesheet" href="//cdnjs.cloudflare.com/ajax/libs/morris.js/0.5.1/morris.css">
 
     <link rel="shortcut icon" href="favicon.ico">
     @yield('css')
+
+    <script src="/js/jquery-1.11.3.min.js"></script>
+    <script src="//cdnjs.cloudflare.com/ajax/libs/raphael/2.1.0/raphael-min.js"></script>
```
Comment (cydrobolt) [inline resources/views/layouts/base.blade.php]: We should download these external JS/CSS libraries locally and access them that way. 
Raphael and Morris specifically could be placed in the CSS & JS sections of the templates re: statistics; we don't need them on `base.blade.php`, because it would unnecessarily slow down the load time.

Also, is there a reason JS, Morris, and Raphael need to be in the `<head>`? Otherwise, we should load them asynchronously at the end of the `<body>` with the other scripts.

File: database/migrations/2016_09_24_095751_create_stats_table.php (hunk 11 above)
Comment (RdN) [inline database/migrations/2016_09_24_095751_create_stats_table.php]: @cydrobolt So it will be backward compatible, but we will have two data sources - clicks column, and clicks table. Ho should be determined which one should be used?

File: database/migrations/2016_09_24_095751_create_stats_table.php (hunk 12 above)
Comment (RdN) [inline database/migrations/2016_09_24_095751_create_stats_table.php]: Added.

File: database/migrations/2016_09_24_095751_create_stats_table.php (hunk 11 above)
Comment (cydrobolt) [inline database/migrations/2016_09_24_095751_create_stats_table.php]: Right. If the environmental variable for additional stats is enabled, then we should use the clicks table for additional data. Otherwise, we should not store data in the clicks table or pull from it.

File: resources/views/layouts/base.blade.php (hunk 13 above)
Comment (jcalonso) [inline resources/views/layouts/base.blade.php]: This assets could be required using npm, and later pre-compiled using elixir/gulp
But probably we should leave this for later, this PR is already quite big.

**Comment by jcalonso** [comment]
I dont see the vendor folder being removed. can we do that please?
Is there anything else holding this PR?

**Comment by cydrobolt** [comment]
We'll remove the vendor folder nonetheless; I'm looking to make some modifications of the graphs with d3.js before I merge this in. Making the "advanced stats" optional is also something we need to fix before integrating this in.

File: resources/views/layouts/base.blade.php (hunk 13 above)
Comment (RdN) [inline resources/views/layouts/base.blade.php]: @cydrobolt Moved to the end of the body. In fact using CDN can speed up the page, depending where it is hosted and where is the user. CDNs usually has edge locations to be able to deliver content as fast as possible. So I suggest to leave it as it is. There are ale other pros of such solution like maximum concurrency connections to same domain - that is why domain sharding is frequently introduced.

**Comment by RdN** [comment]
@jcalonso Totally agree, fighting with vendors files conflicts is hell :)

File: resources/views/layouts/base.blade.php (hunk 13 above)
Comment (RdN) [inline resources/views/layouts/base.blade.php]: @cydrobolt I moved it, but I don't agree. It is same situation as with vendors.

Comment (cydrobolt) [inline resources/views/layouts/base.blade.php]: We've removed the vendor files from the repo :)

**Comment by jcalonso** [comment]
My biggest concern with the pull request is the fact that now there is an extra step on the installation process and also coupling with the Maxmind geo database (and no way to disable this).

I think we should try to keep the project simple for new comers to install, but easy to configure for advanced users or users with more specific requirements. 

A nice option, would be to make the click tracking optional to avoid that extra step of downloading the database. Another option is to fetch the database automatically as part of the setup process.

An event better approach would be to decouple the implementation of ip to geo resolution of Maxmind db by using a package like  http://lyften.com/projects/laravel-geoip/ which offers multiple options to do the ip to geo resolution, by default uses a free rest api, and if required, can be replaced for the Maxmind db. This makes the installation of Polr friction-less to new uses.

Let me know what you think.

**Comment by cydrobolt** [comment]
@jcalonso I agree. There should be as small of an installation learning curve as possible to make it easy for newcomers to configure and use the project. However, I also agree that optional extra stats are extremely useful for enterprise users and other business users that require extra analytics. Without more advanced stats, it would be more difficult to provide the data that some users may expect from a URL shortening platform.

`laravel-geoip` looks very promising. I'll take a look at the package and refactor the code to help make it simpler for users to enable the feature but not overly difficult for new users to get started.

**Comment by rmrhz** [comment]
Was there already an API for Maxmind GeoIP?

**Comment by cydrobolt** [comment]
Closing in favor of #279
